File: server/src/services/gemini.service.js
const { GoogleGenerativeAI, HarmCategory, HarmBlockThreshold } = require('@google/generative-ai');

class GeminiService {
    constructor() {
        const apiKey = process.env.GEMINI_API_KEY;
        if (!apiKey) {
            console.error('Warning: GEMINI_API_KEY is not set in .env');
        }

        this.genAI = new GoogleGenerativeAI(apiKey);
        this.model = this.genAI.getGenerativeModel({
            model: "gemini-2.5-flash",
            generationConfig: {
                temperature: 0.7,
                maxOutputTokens: 1024,
            },
            safetySettings: [
                { category: HarmCategory.HARM_CATEGORY_HARASSMENT, threshold: HarmBlockThreshold.BLOCK_MEDIUM_AND_ABOVE },
                { category: HarmCategory.HARM_CATEGORY_HATE_SPEECH, threshold: HarmBlockThreshold.BLOCK_MEDIUM_AND_ABOVE },
                { category: HarmCategory.HARM_CATEGORY_SEXUALLY_EXPLICIT, threshold: HarmBlockThreshold.BLOCK_MEDIUM_AND_ABOVE },
                { category: HarmCategory.HARM_CATEGORY_DANGEROUS_CONTENT, threshold: HarmBlockThreshold.BLOCK_MEDIUM_AND_ABOVE },
            ]
        });
    }

    async generateAssessment(data) {
        const { symptoms, age, gender, description, language } = data;

        // Map language
        const languageMap = {
            'en': 'English',
            'hi': 'Hindi (हिंदी)',
            'mr': 'Marathi (मराठी)',
            'English': 'English',
            'Hindi': 'Hindi (हिंदी)',
            'Marathi': 'Marathi (मराठी)'
        };

        const targetLanguage = languageMap[language] || languageMap['en'];
        const symptomsText = Array.isArray(symptoms) ? symptoms.join(', ') : symptoms;

        const prompt = `
You are a culturally sensitive health advisor for rural India. Generate personalized health recommendations based on:

Patient Details:
- Gender: ${gender}
- Age: ${age}
- Main Symptoms: ${symptomsText}
- Description: "${description || 'None provided'}"
- Language: ${targetLanguage}

Provide specific guidance in the following JSON format structure (do not use markdown code blocks, just raw JSON):
{
  "summary": "A brief summary (2-3 sentences) explaining the condition in simple terms.",
  "recommendations": "List of 5-7 practical, actionable recommendations. Use bullet points or numbered list in the string.",
  "culturalTips": "Cultural considerations (home remedies, dietary advice common in Indian households).",
  "warningSigns": "When to seek immediate medical attention or call 108.",
  "riskLevel": "Low Risk | Medium Risk | High Risk"
}

IMPORTANT:
- Write ENTIRELY in ${targetLanguage}.
- Use simple, clear language that rural populations can understand.
- Be empathetic and reassuring.
- Do NOT provide a medical diagnosis. Use phrases like "It appears to be...", "Possible causes include...".
- For "riskLevel", estimate based on symptoms (e.g., chest pain = High Risk, mild cold = Low Risk).
`;

        try {
            const result = await this.model.generateContent(prompt);
            const response = await result.response;
            const text = response.text();
            // 1. Clean Markdown Code Blocks
            let cleanedText = text.replace(/```json/gi, '').replace(/```/g, '').trim();

            // 2. Try Standard JSON Parse from Cleaner text
            const jsonMatch = cleanedText.match(/\{[\s\S]*\}/);

            let parsedData = null;

            if (jsonMatch) {
                try {
                    parsedData = JSON.parse(jsonMatch[0]);
                } catch (e) {
                    console.error('Failed to parse (even cleaned) JSON, using manual extractor:', e);
                }
            }

            // 3. Manual Extraction (if parse failed)
            if (!parsedData) {
                const extractField = (fieldName) => {
                    const patterns = [
                        new RegExp(`"${fieldName}"\\s*:\\s*"([^"]*)"`, 'i'),
                        new RegExp(`"${fieldName}"\\s*:\\s*'([^']*)'`, 'i'),
                        new RegExp(`'${fieldName}'\\s*:\\s*"([^"]*)"`, 'i'),
                        new RegExp(`'${fieldName}'\\s*:\\s*'([^']*)'`, 'i'),
                        new RegExp(`${fieldName}\\s*:\\s*"([^"]*)"`, 'i')
                    ];
                    for (const p of patterns) {
                        const m = text.match(p); // Use original text for regex matching to be safe
                        if (m && m[1]) return m[1];
                    }
                    return "";
                };

                const summary = extractField('summary');
                const recommendations = extractField('recommendations');

                // If regex found something, populate parsedData
                if (summary.length > 5 || recommendations.length > 5) {
                    parsedData = {
                        summary,
                        recommendations,
                        warningSigns: extractField('warningSigns'),
                        culturalTips: extractField('culturalTips'),
                        riskLevel: extractField('riskLevel')
                    };
                }
            }

            // 4. Force Normalization (The "Layer")
            if (!parsedData) {
                // FAILURE SCENARIO: JSON parse failed, Regex extraction failed.
                // We likely have a malformed JSON string like: { "summary": "Good...
                // We must STRIP all JSON syntax to save the user experience.

                let safeText = cleanedText;

                // A. Strip wrapping braces
                safeText = safeText.replace(/^[\s\n]*\{/, '').replace(/\}[\s\n]*$/, '');

                // B. Remove common keys if they appear at start of lines (e.g. "summary":)
                const keysToRemove = ['summary', 'recommendations', 'culturalTips', 'warningSigns', 'riskLevel', 'data'];
                keysToRemove.forEach(key => {
                    // Regex removes "key": or 'key': or key:
                    safeText = safeText.replace(new RegExp(`['"]?${key}['"]?\\s*:\\s*`, 'gi'), '');
                });

                // C. Cleanup stray quotes and commas at end of lines
                safeText = safeText.replace(/",\s*$/gm, '\n') // Quote-comma at EOL -> Newline
                    .replace(/',\s*$/gm, '\n')
                    .replace(/"\s*$/gm, '')
                    .replace(/^"/gm, '');      // Quote at start of line

                // D. Ensure we don't return an empty string
                if (safeText.trim().length < 5) {
                    safeText = "Guidance generated, but format was unclear. Please consult a doctor for advice.";
                }

                parsedData = {
                    summary: safeText.trim(),
                    recommendations: [],
                    culturalTips: "",
                    warningSigns: "",
                    riskLevel: "Medium Risk",
                    rawText: safeText.trim(),
                    structuredData: false
                };
            }

            // Final Polish: Ensure types are correct AND Sanitized
            const sanitizeString = (str) => {
                if (typeof str !== 'string') return "";
                // If it STILL looks like JSON code (starts with quotes or braces), nukem.
                if (str.trim().startsWith('{') || str.trim().startsWith('[') || str.trim().startsWith('"')) {
                    return str.replace(/[{}"\[\]]/g, '');
                }
                return str;
            };

            return {
                summary: sanitizeString(parsedData.summary || "Health Guidance"),
                recommendations: Array.isArray(parsedData.recommendations)
                    ? parsedData.recommendations
                    : (parsedData.recommendations ? [sanitizeString(parsedData.recommendations)] : []),
                culturalTips: sanitizeString(parsedData.culturalTips || ""),
                warningSigns: sanitizeString(parsedData.warningSigns || ""),
                riskLevel: sanitizeString(parsedData.riskLevel || "Medium Risk"),
                structuredData: parsedData.structuredData !== false
            };
        } catch (error) {
            console.error('Gemini API Error:', error);
            // Even on API error, return a valid response to prevent client hang
            return {
                summary: "Service temporarily unavailable. Please try again.",
                riskLevel: "Unknown",
                error: true
            };
        }
    }
}

module.exports = new GeminiService();

File: lib/services/gemini_service.dart
import 'package:google_generative_ai/google_generative_ai.dart';

class GeminiService {
  static final GeminiService _instance = GeminiService._internal();
  GenerativeModel? _model;
  static const String apiKey = String.fromEnvironment('GEMINI_API_KEY');

  factory GeminiService() => _instance;

  GeminiService._internal() {
    _initializeService();
  }

  void _initializeService() {
    if (apiKey.isEmpty) {
      print('Warning: GEMINI_API_KEY is missing. Using mock data.');
      return;
    }

    _model = GenerativeModel(
      model: 'gemini-2.5-flash',
      apiKey: apiKey,
      generationConfig: GenerationConfig(
        temperature: 0.7,
        maxOutputTokens: 1024,
      ),
      safetySettings: [
        SafetySetting(HarmCategory.harassment, HarmBlockThreshold.medium),
        SafetySetting(HarmCategory.hateSpeech, HarmBlockThreshold.medium),
        SafetySetting(HarmCategory.sexuallyExplicit, HarmBlockThreshold.medium),
        SafetySetting(HarmCategory.dangerousContent, HarmBlockThreshold.medium),
      ],
    );
  }

  Future<Map<String, String>> generateHealthRecommendations({
    required List<String> symptoms,
    required String riskLevel,
    required String language,
  }) async {
    // Return mock data if service is not initialized (missing API key)
    if (_model == null) {
      await Future.delayed(const Duration(seconds: 1)); // Simulate network delay
      return _getMockData(language);
    }

    try {
      final languageMap = {
        'English': 'English',
        'Hindi': 'Hindi (हिंदी)',
        'Marathi': 'Marathi (मराठी)',
      };

      final targetLanguage = languageMap[language] ?? 'English';
      final symptomsText = symptoms.join(', ');

      final prompt =
      '''
You are a culturally sensitive health advisor for rural India. Generate personalized health recommendations based on:

Symptoms: $symptomsText
Risk Level: $riskLevel
Language: $targetLanguage

Provide recommendations in the following format:
1. A brief summary (2-3 sentences) explaining the condition in simple terms
2. 5-7 practical, actionable recommendations suitable for rural settings
3. Cultural considerations (home remedies, dietary advice common in Indian households)
4. When to seek immediate medical attention

IMPORTANT:
- Write ENTIRELY in $targetLanguage
- Use simple, clear language that rural populations can understand
- Include culturally appropriate advice (e.g., mention tulsi, ginger, turmeric for home remedies)
- Be empathetic and reassuring
- For High Risk: emphasize urgency but avoid panic
- For Medium Risk: balance caution with practical monitoring advice
- For Low Risk: provide comfort while encouraging basic self-care

Format your response as:
SUMMARY: [your summary here]
RECOMMENDATIONS:
- [recommendation 1]
- [recommendation 2]
...
CULTURAL_TIPS: [cultural advice]
WARNING_SIGNS: [when to seek help]
''';

      final content = [Content.text(prompt)];
      final response = await _model!.generateContent(content);

      if (response.text == null || response.text!.isEmpty) {
        throw GeminiException('No response generated from Gemini API');
      }

      return _parseResponse(response.text!);
    } catch (e) {
      if (e is GenerativeAIException) {
        throw GeminiException('Gemini API Error: ${e.message}');
      }
      throw GeminiException('Failed to generate recommendations: $e');
    }
  }

  /// Generates general guidance for "Tell Us More" screen
  Future<Map<String, String>> generateGeneralGuidance({
    required String gender,
    required String age,
    required String description,
    required List<String> selectedChips,
    String language = 'English',
  }) async {
    if (_model == null) {
      await Future.delayed(const Duration(seconds: 1));
      return _getMockData(language);
    }

    try {
      // Basic language map (can be expanded)
      final languageMap = {
        'English': 'English',
        'Hindi': 'Hindi',
        'Marathi': 'Marathi',
      };
      final targetLanguage = languageMap[language] ?? 'English';

      final prompt = '''
You are a preliminary health guide for rural India. A user has reported a problem.
User Details:
- Gender: $gender
- Age: $age
- Additional Symptoms: ${selectedChips.join(', ')}
- Problem Description: "$description"

Provide safe, non-diagnostic guidance in $targetLanguage.
Focus on:
1. Reassurance and calm advice.
2. Basic home care or immediate steps.
3. Clear warning signs indicating when to see a doctor or call 108 (Emergency).

IMPORTANT:
- Do NOT provide a medical diagnosis.
- Use simple, easy-to-understand language.
- Structure with clear headings.

Format your response strictly as:
SUMMARY: [Reassurance and brief understanding of the issue]
ADVICE: 
- [Step 1]
- [Step 2]
WARNING_SIGNS: [Specific signs to go to hospital immediately]
''';

      final content = [Content.text(prompt)];
      final response = await _model!.generateContent(content);

      if (response.text == null || response.text!.isEmpty) {
         throw GeminiException('No response generated.');
      }

      return _parseGeneralResponse(response.text!);

    } catch (e) {
       if (e is GenerativeAIException) {
        throw GeminiException('Gemini API Error: ${e.message}');
      }
      throw GeminiException('Failed to generate guidance: $e');
    }
  }

  Map<String, String> _parseGeneralResponse(String text) {
     final result = <String, String>{};
     // Simple parsing based on the requested format
     // Using regex similar to _parseResponse but adapted for the new keys
     
     // Summary
     var match = RegExp(r'SUMMARY:\s*(.+?)(?=\n\n|ADVICE:)', dotAll: true).firstMatch(text);
     result['summary'] = match?.group(1)?.trim() ?? '';

     // Advice
     match = RegExp(r'ADVICE:\s*(.+?)(?=\n\n|WARNING_SIGNS:)', dotAll: true).firstMatch(text);
     result['advice'] = match?.group(1)?.trim() ?? '';

     // Warning Signs
     match = RegExp(r'WARNING_SIGNS:\s*(.+?)$', dotAll: true).firstMatch(text);
     result['warningSigns'] = match?.group(1)?.trim() ?? '';

     // Fallback
     if (result['summary']!.isEmpty) {
       result['summary'] = text; // Just show raw text if parse fails
     }
     
     return result;
  }

  Map<String, String> _getMockData(String language) {
    if (language == 'Hindi') {
      return {
        'summary': 'आपके लक्षणों के आधार पर, यह एक सामान्य संक्रमण लग रहा है। चिंता की कोई बात नहीं है, लेकिन आराम करना महत्वपूर्ण है।',
        'recommendations': '- खूब पानी और तरल पदार्थ पिएं।\n- पर्याप्त आराम करें।\n- हल्का और सुपाच्य भोजन लें।\n- बुखार होने पर ठंडे पानी की पट्टियां रखें।',
        'culturalTips': 'तुलसी और अदरक की चाय पीने से राहत मिल सकती है।',
        'warningSigns': 'यदि बुखार 3 दिनों से अधिक रहता है या साँस लेने में तकलीफ होती है, तो तुरंत डॉक्टर से संपर्क करें।',
      };
    } else if (language == 'Marathi') {
      return {
        'summary': 'तुमच्या लक्षणांवरून हे सामान्य संसर्ग असल्याचे दिसते. काळजी करण्याचे कारण नाही, पण आराम करणे महत्त्वाचे आहे.',
        'recommendations': '- भरपूर पाणी आणि द्रव पदार्थ प्या.\n- पुरेशी झोप घ्या.\n- हलके आणि पचायला सोपे अन्न खा.\n- ताप असल्यास कोमट पाण्याने अंग पुसून घ्या.',
        'culturalTips': 'तुळस आणि आले टाकलेला चहा प्यायल्याने आराम मिळू शकतो.',
        'warningSigns': 'जर ताप ३ दिवसांपेक्षा जास्त राहिला किंवा श्वास घेण्यास त्रास होत असेल, तर त्वरित डॉक्टरांचा सल्ला घ्या.',
      };
    } else {
      return {
        'summary': 'Based on your symptoms, this appears to be a mild viral infection. There is usually no cause for alarm, but rest is key to recovery.',
        'recommendations': '- Drink plenty of water and fluids.\n- Get adequate rest and sleep.\n- Eat light, home-cooked meals.\n- Monitor your temperature regularly.',
        'culturalTips': 'Drinking warm turmeric milk (Haldi Doodh) before bed can boost immunity.',
        'warningSigns': 'Seek immediate medical help if you experience difficulty breathing, chest pain, or high fever persisting for more than 3 days.',
      };
    }
  }

  Map<String, String> _parseResponse(String responseText) {
    final result = <String, String>{};

    try {
      // Extract summary
      final summaryMatch = RegExp(
        r'SUMMARY:\s*(.+?)(?=\n\n|RECOMMENDATIONS:)',
        dotAll: true,
      ).firstMatch(responseText);
      result['summary'] = summaryMatch?.group(1)?.trim() ?? '';

      // Extract recommendations
      final recommendationsMatch = RegExp(
        r'RECOMMENDATIONS:\s*(.+?)(?=\n\n|CULTURAL_TIPS:|WARNING_SIGNS:|$)',
        dotAll: true,
      ).firstMatch(responseText);
      result['recommendations'] = recommendationsMatch?.group(1)?.trim() ?? '';

      // Extract cultural tips
      final culturalMatch = RegExp(
        r'CULTURAL_TIPS:\s*(.+?)(?=\n\n|WARNING_SIGNS:|$)',
        dotAll: true,
      ).firstMatch(responseText);
      result['culturalTips'] = culturalMatch?.group(1)?.trim() ?? '';

      // Extract warning signs
      final warningMatch = RegExp(
        r'WARNING_SIGNS:\s*(.+?)$',
        dotAll: true,
      ).firstMatch(responseText);
      result['warningSigns'] = warningMatch?.group(1)?.trim() ?? '';

      // If parsing fails, return the full response as summary
      if (result['summary']!.isEmpty && result['recommendations']!.isEmpty) {
        result['summary'] = responseText;
      }

      return result;
    } catch (e) {
      // Fallback: return full response
      return {
        'summary': responseText,
        'recommendations': '',
        'culturalTips': '',
        'warningSigns': '',
      };
    }
  }

  void dispose() {
    // Cleanup if needed
  }
}

class GeminiException implements Exception {
  final String message;

  GeminiException(this.message);

  @override
  String toString() => 'GeminiException: $message';
}
